### PR TITLE
Able to complete the training after these changes

### PR DESCRIPTION
**Pull request by Aryal007**
**ml_classifier** 
- hyperparameter search for best parameters in MLP and added those

**frame.py** 
- fixed output dir
- made loss a class attribute
- changed function name loss -> calc_loss
- changed permute in y value (Wasn't able to run without fixing this)
- infer function returns _(Batch * Channels * Height * Weight)_
- calculate metrics was calculating twice, fixed

**mask.py**
- changed path object to string for concatenation
- changed string to rasterio.crs.CRS objects for comparison
- clipped glaciers to borders.shp (This is needed. I will explain this when you have time)

**post_process.py**
- added default root

**post_process_funs.py**
- create file if not exists
- filter by adding img_mean

**slice.py**
- added img_mean inn json file
- changed soon filename for consistency across files

**train.py**
- batch wise loss visualization using training
- changed path to save_directory

**unet.py**
- changed to 16 channels on first conv_layer

### CODE REVIEW

File: src/slice.py [hunk 1]
```diff
@@ -115,5 +117,5 @@ def wrapper(row):
     para = Parallel(n_jobs=args.n_cpu)
     metadata = para(delayed(wrapper)(k) for k in range(len(paths)))
     metadata = pd.concat(metadata, axis=0)
-    out_path = Path(args.output_dir, f"slices_{args.start_line}-{args.end_line}.geojson")
+    out_path = Path(args.output_dir, f"slices_subset.geojson")
```
Comment (krisrs1128) [inline src/slice.py]: Won't this overwrite different slices, if you run it in parallel?

File: src/frame.py [hunk 2]
```diff
@@ -4,36 +4,39 @@
 import src.metrics
 import src.models.unet
 import torch
+# from torch import optim
```
Comment (krisrs1128) [inline src/frame.py]: I'd prefer if you just remove, rather than comment out. We can always add back if we want later.

File: src/mask.py [hunk 3]
```diff
@@ -36,14 +37,27 @@ def generate_masks(img_paths, shps_paths, output_base="mask",
...
         for path in shps_paths[k]:
-            shps.append(gpd.read_file(path).to_crs(img.meta["crs"].data))
+            gdf = gpd.read_file(path)
+            gdf_crs = rasterio.crs.CRS(gdf.crs)
+            if gdf_crs != img.meta["crs"]:
+                gdf = gdf.to_crs(img.meta["crs"].data)
+            shps.append(gdf)
+
+        if rasterio.crs.CRS(img.meta["crs"].data) != (rasterio.crs.CRS(shps[0].crs) == rasterio.crs.CRS(shps[1].crs)):
+            print("\nImageCRS: ",rasterio.crs.CRS(img.meta["crs"].data))
```
Comment (krisrs1128) [inline src/mask.py:58]: This is smart.

File: src/mask.py [hunk 4]
```diff
@@ -36,14 +37,27 @@ def generate_masks(img_paths, shps_paths, output_base="mask",
...
+    if not os.path.exists(out_dir):
+      os.makedirs(out_dir)
     metadata_path = pathlib.Path(out_dir, "metadata.csv")
     metadata.to_csv(metadata_path, index=False)
 
     def wrapper(k):
         print(f"working on image {k} / {len(img_paths)}")
         img, shps = rasterio.open(img_paths[k]), []
+        
+        # print(shps_paths)
```
Comment (krisrs1128) [inline src/mask.py]: Same as before -- there are a few points in here where things are commented out, where I'd rather they just be removed. As a general rule, I try to make code as  compact as possible (while maintaining clarity).

File: src/postprocess_funs.py [hunk 5]
```diff
@@ -47,7 +48,7 @@ def reshuffle(split_ids, output_dir="output/", n_cpu=3):
         def wrapper(i):
             cur_locs = {}
             for im_type in ["img", "mask"]:
-                print(f"shuffling image {i} - {im_type}")
+                # print(f"shuffling image {i} - {im_type}")
```
Comment (krisrs1128) [inline src/postprocess_funs.py]: same as before, if we don't need something, just delete

File: src/train.py [hunk 6]
```diff
@@ -19,65 +19,73 @@
...
+optim_opts = addict.Dict({"name": "Adam", "args": {"lr": 1e-5}})
+metrics_opts = addict.Dict({"precision": {"threshold": 0.3}, "IoU": {"threshold": 0.3}})
 
+frame = Framework(model_opts=model_opts, optimizer_opts=optim_opts, metrics_opts=metrics_opts, out_dir=Path(path, "models"))
 
 writer = SummaryWriter()
 
 ## Train Loop
-epochs=10
+epochs=1000
```
Comment (krisrs1128) [inline src/train.py]: shouldn't this be read in from one of the configuration files?

**Comment by Aryal007** [comment]
**frame.py** 
- valid for model_opts.name "Unet" and "UnetDropout"
- fixed calculate metrics 

**train.py** 
- added training prediction/validation prediction to tensorboard
- saving input configurations as json in tensorboard
- tensorboard saves train/validation loss and metrics
- prettier commandline train runs
- folder structure consistency with previous steps in training pipeline, change the default value of --path in argument parser after pulling.

**unet_dropout.py**
- experimentation capability with spatial dropout added

**Review by Aryal007 — COMMENTED** [review]
Hi Kris. I made the changes according to your comments. Sorry it took so long
